### PR TITLE
Add ctc_greedy_decoder op

### PR DESCRIPTION
**Pull request by hengzi**
Performs greedy decoding on the logits given in input (best path).
https://tensorflow.google.cn/api_docs/python/tf/nn/ctc_greedy_decoder?hl=zh-cn

### CODE REVIEW

File: oneflow/user/kernels/ctc_greedy_decoder.cu [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+__global__ void CtcGreedyDecodeGpu(int64_t* decoded_ptr, T* neg_sum_logits_ptr,
+                                   const T* log_probs_ptr, const int64_t* input_lengths_ptr,
+                                   const bool merge_repeated, const int64_t max_input_length,
+                                   const int64_t batch_size, const int64_t num_labels) {
+  for (int64_t b = 0; b < batch_size; b++) {
+    if (input_lengths_ptr[b] > max_input_length) __trap();
+  }
+  NdIndexOffsetHelper<int64_t, 3> input_helper(max_input_length, batch_size, num_labels);
+
+  CUDA_1D_KERNEL_LOOP(b, batch_size) {
```
Comment (Ldpe2G) [inline oneflow/user/kernels/ctc_greedy_decoder.cu:35]: 你这里是启动 batch_size / kCudaThreadsNumPerBlock 个block，block内每个 thread，处理一个序列的decode计算过程，感觉这样子并行度有点低。

我有个想法，你可以试试，就是你可以改为启动 batch 个block，每个block的线程协同处理一个序列的decode(一个batch)计算。具体可以这样，一开始每个 thread 负责处理序列一个点的 max_value 和 max_index 的计算，然后把结果存到  shared memory，然后block内每个线程计算完之后，同步一下。最后再由thread 0 计算 neg_sum_logits 和 decoded 的内容。

Comment (hengzi) [reply oneflow/user/kernels/ctc_greedy_decoder.cu:35]: ok 我试着优化一下

File: oneflow/user/kernels/ctc_greedy_decoder.cu [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  }
+}
+template<typename T>
+struct CTCGreedyDecoderFunctor<DeviceType::kGPU, T> final {
+  void operator()(DeviceCtx* ctx, int64_t* decoded_ptr, T* neg_sum_logits_ptr,
+                  const T* log_probs_ptr, const int64_t* input_lengths_ptr,
+                  const bool merge_repeated, const int64_t max_input_length,
+                  const int64_t batch_size, const int64_t num_labels) {
+    int32_t thread_num = batch_size * kCudaThreadsNumPerBlock;
+    int64_t shared_mem_size = max_input_length * sizeof(int64_t);
```
Comment (hengzi) [inline oneflow/user/kernels/ctc_greedy_decoder.cu:82]: 我测试了下，shared memory 大于 6144*8就无法启动kernel，这个大小是否满足需求

Comment (Ldpe2G) [reply oneflow/user/kernels/ctc_greedy_decoder.cu:82]: 那就加个分支判断吧，如果小于这个大小就走这个分支，大于就用原来naive的实现

File: oneflow/user/kernels/ctc_greedy_decoder.cu [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    for (int64_t t = tid; t < max_input_length; t += blockDim.x) {
+      shared_max_indices[t] = -1;
+      T max_value = -FLT_MAX;
+      const T* prob_data_t = &log_probs_ptr[input_helper.NdIndexToOffset(t, b, 0)];
+      FOR_RANGE(int64_t, c, 0, num_labels) {
+        if (prob_data_t[c] > max_value) {
+          shared_max_indices[t] = c;
+          max_value = prob_data_t[c];
+        }
+      }
```
Comment (Ldpe2G) [inline oneflow/user/kernels/ctc_greedy_decoder.cu:55]:
```C++
int64_t max_idx = 0;
FOR_RANGE(int64_t, c, 0, num_labels) {
     cosnt T prob = prob_data_t[c];
     if (prob > max_value) {
          max_idx = c;
          max_value = prob;
     }
}
shared_max_indices[t] = max_idx;
```
感觉 50 行这个循环改成这样好些

File: oneflow/user/kernels/ctc_greedy_decoder.cu [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    __syncthreads();
+
+    if (tid == 0) {
+      int64_t prev_indices = -1, t_dec = 0;
+      FOR_RANGE(int64_t, t, 0, input_lengths_ptr[b]) {
+        const T* prob_data_t = &log_probs_ptr[input_helper.NdIndexToOffset(t, b, 0)];
+
```
Comment (Ldpe2G) [inline oneflow/user/kernels/ctc_greedy_decoder.cu:64]: 下面 65 ~ 71 行多次用到了 `shared_max_indices[t]`，可以先用个局部变量存着，不用反复读

**Review by Ldpe2G — APPROVED** [review]
